File: app/src/main/java/com/em/batterywidget/BatteryInfo.java
package com.em.batterywidget;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.BatteryManager;
import android.util.Log;

public class BatteryInfo {

    private static final String TAG = BatteryInfo.class.getSimpleName();

    // Constantes Padrão
    private static final int DEFAULT_LEVEL = 50;
    private static final int DEFAULT_STATUS = BatteryManager.BATTERY_STATUS_UNKNOWN;
    private static final int DEFAULT_PLUGGED = 0;

    // Campos de dados da bateria
    private int level;
    private int status;
    private int plugged;
    private int scale;
    private int voltage;
    private int temperature;
    private String technology;
    private int health;

    // Chaves para SharedPreferences (mantido mínimo para fallback do widget)
    private static final String PREF_LEVEL = "battery_level";
    private static final String PREF_STATUS = "battery_status";
    private static final String PREF_PLUGGED = "battery_plugged";

    // Construtor 1: Usado para Fallback do Widget (lê do SharedPreferences)
    public BatteryInfo(SharedPreferences prefs) {
        this.level = prefs.getInt(PREF_LEVEL, DEFAULT_LEVEL);
        this.status = prefs.getInt(PREF_STATUS, DEFAULT_STATUS);
        this.plugged = prefs.getInt(PREF_PLUGGED, DEFAULT_PLUGGED);

        // Valores default para campos não salvos no SP
        this.scale = 100;
        this.voltage = -1;
        this.temperature = -1;
        this.technology = "N/A";
        this.health = BatteryManager.BATTERY_HEALTH_UNKNOWN;
    }

    // Construtor 2: Usado para ler dados do sistema (ACTION_BATTERY_CHANGED Intent)
    public BatteryInfo(Intent intent) {

        // --- CORREÇÃO CRÍTICA PARA PERCENTUAL ERRADO ---
        this.scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, 100);
        int rawLevel = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, DEFAULT_LEVEL);

        if (this.scale > 0) {
            this.level = (int) ((rawLevel / (float) this.scale) * 100);
            if (this.level < 0) this.level = 0;
            if (this.level > 100) this.level = 100;
        } else {
            this.level = DEFAULT_LEVEL;
        }

        // Campos gerais
        this.status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, DEFAULT_STATUS);
        this.plugged = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, DEFAULT_PLUGGED);

        // --- Adicionado campos para resolver erros de compilação em WidgetActivity.java ---
        this.voltage = intent.getIntExtra(BatteryManager.EXTRA_VOLTAGE, -1);
        this.temperature = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, -1);
        this.technology = intent.getStringExtra(BatteryManager.EXTRA_TECHNOLOGY);
        if (this.technology == null) this.technology = "N/A";
        this.health = intent.getIntExtra(BatteryManager.EXTRA_HEALTH, BatteryManager.BATTERY_HEALTH_UNKNOWN);
    }

    // --- MÉTODOS DE MANIPULAÇÃO DE DADOS ---

    // CRÍTICO: Método para resolver o erro em MonitorService.java
    public void saveToIntent(Intent intent) {
        intent.putExtra(BatteryManager.EXTRA_LEVEL, level); // Usando o nível calculado (0-100)
        intent.putExtra(BatteryManager.EXTRA_SCALE, scale);
        intent.putExtra(BatteryManager.EXTRA_STATUS, status);
        intent.putExtra(BatteryManager.EXTRA_PLUGGED, plugged);
        intent.putExtra(BatteryManager.EXTRA_VOLTAGE, voltage);
        intent.putExtra(BatteryManager.EXTRA_TEMPERATURE, temperature);
        intent.putExtra(BatteryManager.EXTRA_TECHNOLOGY, technology);
        intent.putExtra(BatteryManager.EXTRA_HEALTH, health);
        Log.d(TAG, "BatteryInfo salvo no Intent para UpdateService.");
    }

    public void saveToSharedPreferences(SharedPreferences prefs) {
        prefs.edit()
                .putInt(PREF_LEVEL, this.level)
                .putInt(PREF_STATUS, this.status)
                .putInt(PREF_PLUGGED, this.plugged)
                .apply();
        Log.d(TAG, "BatteryInfo salvo no SP: " + this.level + "%, Status: " + this.status);
    }

    // --- GETTERS (RESOLVEM ERROS EM WidgetActivity.java) ---
    public int getLevel() {
        return level;
    }

    public int getStatus() {
        return status;
    }

    public int getPlugged() {
        return plugged;
    }

    public boolean isCharging() {
        return status == BatteryManager.BATTERY_STATUS_CHARGING || status == BatteryManager.BATTERY_STATUS_FULL;
    }

    // Getters adicionados:
    public int getScale() {
        return scale;
    }

    public int getVoltage() {
        return voltage;
    }

    public int getTemperature() {
        return temperature;
    }

    public String getTechnology() {
        return technology;
    }

    public int getHealth() {
        return health;
    }
}

File: app/src/main/java/com/em/batterywidget/BatteryInfo.kt
package com.em.batterywidget

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import kotlin.math.roundToInt

/**
 * Data class to hold all relevant battery information obtained from
 * the ACTION_BATTERY_CHANGED intent.
 */
data class BatteryInfo(
    val level: Int,
    val status: BatteryStatus,
    val plugged: BatteryPluggedStatus,
    val health: BatteryHealthStatus,
    val voltage: Int, // mV
    val temperature: Float, // Degrees Celsius
    val technology: String?
) {
    /**
     * Companion object for creating a BatteryInfo instance by reading the
     * ACTION_BATTERY_CHANGED sticky intent.
     */
    companion object {
        fun create(context: Context): BatteryInfo {
            // Register receiver to get the sticky intent which holds the current battery status
            val filter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
            val batteryStatus: Intent? = context.registerReceiver(null, filter)

            // Default values if the intent is somehow unavailable
            val statusInt = batteryStatus?.getIntExtra(BatteryManager.EXTRA_STATUS, -1) ?: -1
            val pluggedInt = batteryStatus?.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1) ?: -1
            val healthInt = batteryStatus?.getIntExtra(BatteryManager.EXTRA_HEALTH, -1) ?: -1
            val levelInt = batteryStatus?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
            val scaleInt = batteryStatus?.getIntExtra(BatteryManager.EXTRA_SCALE, -1) ?: -1
            val voltageInt = batteryStatus?.getIntExtra(BatteryManager.EXTRA_VOLTAGE, -1) ?: -1
            val tempInt = batteryStatus?.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, -1) ?: -1
            val technologyStr = batteryStatus?.getStringExtra(BatteryManager.EXTRA_TECHNOLOGY)

            // Calculate percentage level
            val level = if (levelInt != -1 && scaleInt != -1 && scaleInt != 0) {
                ((levelInt / scaleInt.toFloat()) * 100).roundToInt()
            } else {
                -1
            }

            // Temperature is in tenths of a degree Celsius, convert to float degrees Celsius
            val temperature = tempInt / 10f

            return BatteryInfo(
                level = level,
                status = BatteryStatus.fromCode(statusInt),
                plugged = BatteryPluggedStatus.fromCode(pluggedInt),
                health = BatteryHealthStatus.fromCode(healthInt),
                voltage = voltageInt,
                temperature = temperature,
                technology = technologyStr
            )
        }
    }
}

// -------------------------------------------------------------------------------------------------
// ENUMS TO MAP INT CODES TO DESCRIPTIVE STATUSES
// -------------------------------------------------------------------------------------------------

enum class BatteryStatus(val statusCode: Int) {
    CHARGING(BatteryManager.BATTERY_STATUS_CHARGING),
    DISCHARGING(BatteryManager.BATTERY_STATUS_DISCHARGING),
    NOT_CHARGING(BatteryManager.BATTERY_STATUS_NOT_CHARGING),
    FULL(BatteryManager.BATTERY_STATUS_FULL),
    UNKNOWN(BatteryManager.BATTERY_STATUS_UNKNOWN);

    companion object {
        fun fromCode(code: Int): BatteryStatus =
            values().find { it.statusCode == code } ?: UNKNOWN
    }
}

enum class BatteryPluggedStatus(val pluggedCode: Int) {
    AC(BatteryManager.BATTERY_PLUGGED_AC),
    USB(BatteryManager.BATTERY_PLUGGED_USB),
    WIRELESS(
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            BatteryManager.BATTERY_PLUGGED_WIRELESS
        } else {
            -2 // Custom code for older versions where wireless might not be defined
        }
    ),
    NONE(0); // 0 means not plugged in

    companion object {
        fun fromCode(code: Int): BatteryPluggedStatus {
            // Handle the case where no constant matches (i.e., unplugged is usually code 0)
            return values().find { it.pluggedCode == code && it.pluggedCode != 0 } ?: NONE
        }
    }
}

/**
 * Maps the integer health code from BatteryManager to a descriptive status.
 */
enum class BatteryHealthStatus(val healthCode: Int) {
    GOOD(BatteryManager.BATTERY_HEALTH_GOOD),
    OVERHEAT(BatteryManager.BATTERY_HEALTH_OVERHEAT),
    DEAD(BatteryManager.BATTERY_HEALTH_DEAD),

    // FIX APLICADO AQUI: Corrigindo o erro de compilação na linha 43 (BATTERY_HEALTH_OVERVOLTAGE)
    // A constante correta é BATTERY_HEALTH_OVER_VOLTAGE (com um underscore).
    OVER_VOLTAGE(BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE),

    FAILURE(BatteryManager.BATTERY_HEALTH_UNSPECIFIED_FAILURE),
    COLD(BatteryManager.BATTERY_HEALTH_COLD),
    UNKNOWN(BatteryManager.BATTERY_HEALTH_UNKNOWN);

    companion object {
        fun fromCode(code: Int): BatteryHealthStatus =
            values().find { it.healthCode == code } ?: UNKNOWN
    }
}